proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java

```diff
@@ -50,6 +58,10 @@ public ConnectedPlayer(GameProfile profile, MinecraftConnection connection, Inet
         this.profile = profile;
         this.connection = connection;
         this.virtualHost = virtualHost;
+
+        // load permissions provider
+        EventManager eventManager = VelocityServer.getServer().getEventManager();
+        this.permissionFunction = eventManager.fire(new PermissionsSetupEvent(this, DEFAULT_PERMISSIONS)).join().createFunction(this);
```

Comment:
I prefer if this did not use `.join()`, this blocks the Netty I/O threads and defeats the point of moving event listeners off the main thread (which was specifically intended to avoid executing or waiting on "alien" code).

Reply:
I don't doubt that LuckPerms would sanely handle this event. But I'm not so sure about other plugins, which might do things like run a SQL query when this event is fired, which was a major consideration in forcing all events to be run "asynchronously" in Velocity. It's still early days, so we could have a discussion to revisit this decision (perhaps there are valid performance considerations).

The way I see it: Velocity should do its very best to ensure the Netty I/O threads are runnable at all times doing useful work like processing packets. Using `join()` here defeats the point: it's blocking waiting on alien code to finish up.

Reply:
I understand the aim, but as I said, I don't see a sane way to achieve it.

1) Call the event and store a `CompletableFuture<PermissionsSetupEvent>` instance on the `ConnectedPlayer`. I dislike this because it's messy, and you're really just moving the problem elsewhere - we'll have to block when hasPermission is eventually called.

2) Init the function lazily when the future completes - causes race conditions, not really an acceptable solution imo.

Reply:
Or simply initialize an `UNDEFINED`-only permission function in `ConnectedPlayer`, fire the event once `ConnectedPlayer` is constructed, wait for it to complete asynchronously, set the new permission function, and allow the normal login process to complete.

This should be ok: plugins handling permission setup should not rely on the player having permissions already.

Reply:
And if the normal login process completes before permissions are loaded?

As you said: 
> But I'm not so sure about other plugins, which might do things like run a SQL query when this event is fired

Which:
> causes race conditions, not really an acceptable solution imo

Reply:
> And if the normal login process completes before permissions are loaded?

I did not mean "the normal login process completes concurrently with the event execution". Simply use `whenCompleteAsync` to get the event execution back on the Netty event loop and finish the login process there.

Reply:
Okay, I've moved the loading into the `LoginSessionHandler`, so everything is loading off the netty threads now. :)